# HG changeset patch
# User Elte Hupkes <elte@hupkes.org>
# Date 1447238941 -3600
#      Wed Nov 11 11:49:01 2015 +0100
# Branch gazebo6-revolve
# Node ID 19c3bb917fc21f08ab412b62b6ebce2ad35e1465
# Parent  6de7a50a8f87adb88f8974b2ec4553a06f67e6ba
Fixing Link space not being deleted when self_collide is enabled

diff -r 6de7a50a8f87 -r 19c3bb917fc2 gazebo/physics/ode/ODELink.cc
--- a/gazebo/physics/ode/ODELink.cc	Tue Nov 10 20:50:00 2015 +0100
+++ b/gazebo/physics/ode/ODELink.cc	Wed Nov 11 11:49:01 2015 +0100
@@ -35,7 +35,8 @@
 
 //////////////////////////////////////////////////
 ODELink::ODELink(EntityPtr _parent)
-    : Link(_parent)
+    : Link(_parent),
+      hasOwnSpace(false)
 {
   this->linkId = NULL;
 }
@@ -46,6 +47,11 @@
   if (this->linkId)
     dBodyDestroy(this->linkId);
   this->linkId = NULL;
+
+  if (this->spaceId && this->hasOwnSpace)
+    dSpaceDestroy(this->spaceId);
+  this->spaceId = NULL;
+  this->hasOwnSpace = false;
 }
 
 //////////////////////////////////////////////////
@@ -182,6 +188,11 @@
     dBodyDestroy(this->linkId);
   this->linkId = NULL;
 
+  if (this->spaceId && this->hasOwnSpace)
+    dSpaceDestroy(this->spaceId);
+  this->spaceId = NULL;
+  this->hasOwnSpace = false;
+
   this->odePhysics.reset();
 }
 
@@ -223,7 +234,11 @@
 {
   this->sdf->GetElement("self_collide")->Set(_collide);
   if (_collide)
-    this->spaceId = dSimpleSpaceCreate(this->odePhysics->GetSpaceId());
+  {
+    // Use method to automatically delete existing space
+    this->SetSpaceId(dSimpleSpaceCreate(this->odePhysics->GetSpaceId()));
+    this->hasOwnSpace = true;
+  }
 }
 
 //////////////////////////////////////////////////
@@ -694,6 +709,12 @@
 //////////////////////////////////////////////////
 void ODELink::SetSpaceId(dSpaceID _spaceid)
 {
+  if (this->spaceId && this->hasOwnSpace)
+  {
+    // Destroy space which was originally created by this link
+    dSpaceDestroy(this->spaceId);
+    this->hasOwnSpace = false;
+  }
   this->spaceId = _spaceid;
 }
 
diff -r 6de7a50a8f87 -r 19c3bb917fc2 gazebo/physics/ode/ODELink.hh
--- a/gazebo/physics/ode/ODELink.hh	Tue Nov 10 20:50:00 2015 +0100
+++ b/gazebo/physics/ode/ODELink.hh	Wed Nov 11 11:49:01 2015 +0100
@@ -185,6 +185,9 @@
       /// \brief Collision space id.
       private: dSpaceID spaceId;
 
+      /// \brief Whether or not this link has create its own space
+      private: bool hasOwnSpace;
+
       /// \brief Cache force applied on body
       private: math::Vector3 force;
 
